Survey Test 개발

truthgarnet opened the pull request:
- Entity 추가 87991fb9cae4fde863aa8debf5079a87ae5c7053
- Service 추가 c2268f4926e1e59345e0a6a2b911e06b975357cd
- Repository 추가 c2268f4926e1e59345e0a6a2b911e06b975357cd, c12e9f2ea69c46f1a207c4539433234063ac4317
- Mapper 추가 9c39ce3daf8adc13a33e1e3445656ecbd16530c3
- Test 코드 추가 f2e5138f90be811c9b58cc69d14fdc0bf573d63a

truthgarnet committed:
chore: add Entities

truthgarnet committed:
feat: add SurveyRepository, SurveyMapper, SurveyService

truthgarnet committed:
feat: add UserRepository, UserService

truthgarnet committed:
feat: add Question, SurveyAnswer, UserAnswer Repositories

truthgarnet committed:
chore: [api] add test dependencies, [domain] add annotationProcessor lombok

truthgarnet committed:
refactor: change DTO,Entity @Builder

truthgarnet committed:
feat: add mapper
-[Survey]: toSurveyResponse, toSurveyFindAll, toSurveyEntity
-[SurveyAnswer]: toSurveyAnswerEntity
-[Question]: toQuestionEntityList

truthgarnet committed:
fix: add setSurveyId

truthgarnet committed:
test: Survey
- getSurveyList: 리스트 조회
- getSurvey: 단건 조회
- addSurvey: 추가
- updateAll: 수정
- deleteSurvey: 삭제

truthgarnet committed:
fix: change to builder pattern at SurveyResponse
- remove duplication code (contentType)
- change body to content

truthgarnet committed:
chore: add AuditingEntityListener
- @CreatedDate
- @LastModifiedDate

truthgarnet committed:
refactor: remove pagination getSurveyList

truthgarnet committed:
chore: [common] add jpa dependency

truthgarnet committed:
fix: missing param at mapper
- createdDate
- updatedDate
- surveyVersion

truthgarnet committed:
test: add reset survey_id

truthgarnet committed:
chore: move wrong position application.yml

truthgarnet committed:
feat: add UpdatePart

f-lab-lyan commented on survey-api/src/main/java/org/kong/survey/mapper/QuestionMapperImpl.java:
```diff
@@ -0,0 +1,30 @@
+            return List.of(); // 빈 리스트 반환
+        }
+
+        return questionRequestList.stream()
```
라이브러리를 쓰는 이유는 믿고 쓰는 것인데, 그 부분에 대해서 좀 더 생각해보시면 좋을 것 같습니다.

f-lab-lyan commented on survey-api/src/test/java/SurveyTest.java:
```diff
@@ -0,0 +1,196 @@
+import org.springframework.transaction.annotation.Transactional;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+
+@SpringBootTest(classes = SurveyApplication.class, webEnvironment = SpringBootTest.WebEnvironment.MOCK)
```
```suggestion
@SpringBootTest(classes = SurveyApplication.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```

f-lab-lyan commented on survey-api/src/test/java/SurveyTest.java:
```diff
@@ -0,0 +1,196 @@
+@Import(JpaConfig.class)
+public class SurveyTest {
+
+    @Autowired
+    private MockMvc mockMvc;
```
```suggestion
    @Autowired
    private TestRestTemplate testRestTemplate;
```

truthgarnet committed:
chore: delete duplication dependencies build.gradle

truthgarnet committed:
fix: add missing @Component at QuestionMapperImpl

truthgarnet committed:
refactor: Change Naming

truthgarnet committed:
chore: add assertJ dependency

truthgarnet committed:
@Todo: develop TestRestTemplate

truthgarnet committed:
refactor: remove dupli dependencies

truthgarnet committed:
fix: @AllArgsConstructor, @NoArgsConstructor

truthgarnet committed:
feat: add addSurvey RestTest

truthgarnet committed:
fix: move surveyRepository.save()
- survey 체크 전에 나오는 것을 후로 위치 이동
- @TODO: Questions save 추가

truthgarnet committed:
feat: add getSurveyList()
- @TODO: Questions 추가 하기

truthgarnet committed:
feat: add getSurvey()
- @TODO: Questions 추가 하기

truthgarnet committed:
test: develop getSurvey(), updateAll(), deleteSurvey()

truthgarnet committed:
feat: add AbstractTestExecutionListener for TRUNCATE Table
- @interface로 다른 테스트에도 쉽게 적용 가능
- AbstractTestExecutionListener 사용으로 테스트의 실행 주기에 개입
- 참고 사이트: https://mangkyu.tistory.com/264

truthgarnet committed:
refactor: divide folder integration, unit

truthgarnet committed:
test: add Unit Test
- getSurvey_Exception
- getSurvey_Success
- getSurveyList_Success
- delete_Success
- updateAll_Success
- addSurvey_Success

truthgarnet committed:
refactor: remove duplication junit

truthgarnet committed:
refactor: QuestionType move to dto -> entity

truthgarnet committed:
feat: add toQuestionResponseList

truthgarnet committed:
chore: add @Setter, @Builder

truthgarnet committed:
chore: add questionType

truthgarnet committed:
feat: add QuestionEntity at service

truthgarnet committed:
test: add QuestionType

truthgarnet committed:
test: add Exception
- not find surveyId

truthgarnet committed:
feat: change mapper MapperImpl to using mapstruct mapper Interface

truthgarnet committed:
feat: add getSurveyList()
- @TODO: fix org.springframework.http.converter.HttpMessageConversionException: Type definition error

f-lab-lyan commented on buildSrc/src/main/groovy/survey.java-conventions.gradle:
```diff
@@ -15,7 +15,13 @@ repositories {
+    compileOnly 'org.projectlombok:lombok:1.18.36'
+    annotationProcessor 'org.projectlombok:lombok:1.18.36'
+
     testImplementation "junit:junit:4.13"
```
junit5로 바꾸는 것이 좋을 것 같습니다.

f-lab-lyan commented on buildSrc/src/main/groovy/survey.java-conventions.gradle:
```diff
@@ -15,7 +15,13 @@ repositories {
 dependencies {
-    implementation("org.springframework.boot:spring-boot-gradle-plugin:3.4.3")
-    implementation("io.spring.gradle:dependency-management-plugin:1.0.11.RELEASE")
+    implementation 'org.springframework.boot:spring-boot-gradle-plugin:3.4.3'
+    implementation 'io.spring.gradle:dependency-management-plugin:1.0.11.RELEASE'
```
플러그인 블럭으로 옮겨야하는 게 아닐까요?

f-lab-lyan commented on survey-api/build.gradle:
```diff
@@ -4,20 +4,19 @@ plugins {
     testImplementation 'org.springframework.boot:spring-boot-testcontainers:3.4.3'
-    testImplementation 'org.junit.jupiter:junit-jupiter:5.8.1'
     testImplementation 'org.assertj:assertj-core:4.0.0-M1'
+    testAnnotationProcessor 'org.mapstruct:mapstruct-processor:1.6.3'
```
https://docs.spring.io/spring-boot/docs/3.1.1/reference/html/dependency-versions.html#appendix.dependency-versions

f-lab-lyan commented on survey-api/src/main/java/org/kong/survey/dto/Question.java:
```diff
@@ -21,13 +19,17 @@ public class Question {
     }
 
     @Getter
+    @Setter
+    @Builder
+    @NoArgsConstructor
```
```suggestion
    @Setter
    @Builder
    @NoArgsConstructor    
```

각각의 어노테이션의 의도에 대해서 얘기해보도록 하죠.

f-lab-lyan commented on survey-api/src/main/java/org/kong/survey/service/SurveyService.java:
```diff
@@ -25,64 +34,115 @@ public class SurveyService {
     @Autowired
     private QuestionRepository questionRepository;
 
+    @Autowired
```
변수 인젝션은 좋은 방식은 아닌 것 같아요.

truthgarnet replied:
private @RequiredArgsConstructor로 변경 완료 했습니다!

f-lab-lyan commented on survey-api/src/main/java/org/kong/survey/service/SurveyService.java:
```diff
@@ -25,64 +34,115 @@ public class SurveyService {
 
     public Survey.Response findBySurveyId(Integer surveyId) {
-        SurveyEntity survey = surveyRepository.findBySurveyId(surveyId).orElseThrow(() -> new RuntimeException());
+        SurveyEntity survey = surveyRepository.findBySurveyId(surveyId).orElseThrow(() -> new RuntimeException("설문지를 찾을 수 없습니다."));
```
```suggestion
        surveyRepository.findBySurveyId(surveyId).map(survey -> {
        }).then.....
```
메소드 체이닝을 사용해 보면 좋을 것 같아요.

truthgarnet replied:
저는 Optional을 사용해서 객체 자체를 null을 확인 후, 뽑아내서 mapper에 넣어서 dto를 만드는 방식을 진행하는 데, map을 사용해서 어떤 이점을 얻을 수 있는 지 잘 모르겠습니다. 제가 개발을 구조를 잘 못 설정했을까요???
만약 SurveyEntity가 `@OneToMany`로 `List<QuestionEntity>`를 가져온다면 surveyRepository.findBySurveyId(surveyId).map(survey::getQuestions).orElseThrow(new RuntimeException());
이런 식으로 짜는 게 코드 가독성으로 좋을 것 같은데, 이 경우에는 메소드 체이닝으로 얻을 수 있게 뭔지 잘 모르겠습니다.

f-lab-lyan commented on survey-api/src/main/java/org/kong/survey/service/SurveyService.java:
```diff
@@ -25,64 +34,115 @@ public class SurveyService {
-        SurveyEntity survey = surveyRepository.findBySurveyId(surveyId).orElseThrow(() -> new RuntimeException());
+        SurveyEntity survey = surveyRepository.findBySurveyId(surveyId).orElseThrow(() -> new RuntimeException("설문지를 찾을 수 없습니다."));
+
+        List<QuestionEntity> questionEntities = questionRepository.findBySurvey_surveyId(surveyId);
+        List<Question.Response> questionResponses = new ArrayList<>();
+        if (questionEntities != null || questionEntities.size() > 0) {
+            for (int i = 0; i < questionEntities.size(); i++) {
```
```suggestion
        List<Question.Response> questionResponses = questionRepository.findBySurvey_surveyId(surveyId).map(question -> {
            return questionMapper.toQuestionResponse(question);
        });
```

f-lab-lyan commented on survey-api/src/main/java/org/kong/survey/service/SurveyService.java:
```diff
@@ -25,64 +34,115 @@ public class SurveyService {
+            log.info("Questions 입장: {}", request.getQuestions());
+            List<Question.Request> questions = request.getQuestions();
+            List<QuestionEntity> changeQuestionEntities = questionMapper.toQuestionEntityList(changeSurvey, questions);
+            for (QuestionEntity question : changeQuestionEntities) {
```
saveAll()이 좋지 않을까요?

f-lab-lyan reviewed (approved):
수고하셨습니다. 코멘트에 대해서는 다음 PR에서 작업하시죠.
